### ISSUE TITLE
Scripting with you-get

### ISSUE THREAD

**Issue by soimort** [soimort/you-get#851]
Some of you have asked me whether it's possible to use `you-get` as a library. Surely you can, in spite of the fact that (currently) there is nothing like an API documentation (on rtfd.org or nowhere).

Since I've been too (busy|lazy) to respond to such emails, I'll shortly have a write-up here. If you have further questions, ask here in this thread; don't send me private emails. (so I don't have to repeat the same answers to each and every similar question frequently asked by different people)

## Use existing extractors of `you-get`

Import the following:

```py
from you_get.common import *
from you_get.util import *
from you_get.extractors import *
```

Suppose you have the `url` (which is a string) of a YouTube page, to extract the video is as simple as:

```py
        yt, yt.url = YouTube(), url
        yt.prepare()
```

Then you will have `yt.streams` and `yt.dash_streams`. `yt.dash_streams` is the more recommended way to extract audio/video data, for it often contains higher quality streams rather than `yt.streams`, but you'll need some basic knowledge on how to use FFmpeg to concatenate these.

(VEVO videos or videos that require decrypted signatures to extract are a little more tricky to implement; for simplicity it's not handled here in this short example)

You might want to sort all the streams by size:

```py
        r = sorted(yt.dash_streams,
                   key=lambda i: yt.dash_streams[i]['size'])
```

Then `yt.dash_streams[r[-1]]['src']` would be a pair of Video/Audio streams with highest possible quality.

With the information given above, you can easily implement a script `yt-oga` that downloads a YouTube video as an Ogg audio file (which can be very useful if you love listening to music on YouTube but don't really need the videos).
The full code can be found here: **[yt-oga](https://github.com/soimort/dotfiles/blob/8862fc4034a2cb8ad2f2c1d3b92164ab74a7d2b8/Scripts/yt-oga)**. (feel free to reuse)

## Implement your own web scraper / downloader

`you_get` provides a bunch of handy library functions, including a shiny download progress bar that you can benefit from if you're a Python newbie or just want to make a quick-and-cheap crawler to download some images / files from the nasty web.

Use `get_content(url, headers)` (`headers` is optional) to fetch the content from a page (usually just some html, xml or json text), and `download_urls(urls, title, ext, total_size, output_dir, headers)` (`output_dir` and `headers` are optional) to actually download anything.

You can, of course, achieve exactly the same thing with Python's [urllib](https://docs.python.org/3/library/urllib.html) or [Requests](http://docs.python-requests.org/en/latest/), but 1) they won't download files to your disk on their own, so you're still in need of some common routines to handle file opening / writing; 2) they don't come with a nice-looking progress bar (because they're not designed as a "downloader" library). With the `you_get` library you'll get all these reusable functions for free, so you can just focus on your regex and REST API logic (which is the interesting part).

An example script: **[logirl](https://github.com/soimort/dotfiles/blob/8862fc4034a2cb8ad2f2c1d3b92164ab74a7d2b8/Scripts/logirl)** (which just scrapes some... images)

We roughly follow the semantic versioning rules, so the current API is unlikely to change during the lifespan of **you-get `0.4.x`**. And, in the future, if anything may break, hopefully the API docs should be finalized by then.

**Comment by wushilin** [comment]
Love this... I have been parsing STDOUT with regex, this would be awesome.

**Comment by wushilin** [comment]
YouKu URL extraction isn't quite friendly. I have to call extract method, but it replied on self.sorted_streams to be injected by an external class (VideoExtractor), If you can give a handy example to extract the URLs for a youku video, it will be great.

<pre>
yk, yk.url = Youku(), url
yk.prepare()
yk.extract() <= this doesn't work because sorted_streams isn't defined

print(yk.streams)
r = sorted(yk.streams, key=lambda i: yk.streams[i]['size'])
print(r)

pp.pprint(yk.streams_parameter)
print(yk.streams[r[-1]])
</pre>

**Comment by soimort** [comment]
Yes, the Youku extractor is a bit quirky (even confusing). I'll be striving to fix this inconsistency in 0.5, but for now, this might be only way doing it (though hackish).

I guess `Youku().download_by_url` or `download_by_vid` could be useful as well.

**Comment by johnsmeeth** [comment]
What is difference between 2 source on src list. ex.
`list_sources = yt.dash_streams['278']['src']
len(list_sources) # 2: what difference?
`
Many thanks your job.

**Comment by soimort** [comment]
@johnsmeeth

DASH videos are streamed in a video/audio-separated manner, so `list_sources` is always a list of 2 (one for audio-only and one for video-only).

**Comment by johnsmeeth** [comment]
Ok, thanks @soimort . Can you example get link facebook video?

**Comment by gengjiawen** [comment]
@soimort will you consider put this in wiki, this will make it more official :)
